Xcode 8 + Swift 3

mohpor opened the pull request.

mohpor commented:
Some notes:

I have tested the framework locally on my computer and again on my devices, ran the test and everything went fine, all tests' passed. Except, when actually testing it on device, the TwitterBot example would crash upon displaying the action sheet, complaining about internal inconsistency with number of rows and the data set.

It's worth taking a look.

m-revetria commented on Source/ActionController.swift:
```diff
@@ -265,340 +266,340 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
+    open func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
         if hasHeader() && section == 0 { return 0 }
-        if settings.behavior.useDynamics && section > _dynamicSectionIndex {
+        if settings.behavior.useDynamics && section > _dynamicSectionIndex ?? 0 {
```
This is not exactly the same, now if `_dynamicSectionIndex` is null and `section` is 0, then this function si returning `_sections[actionSectionIndexFor(section)].actions.count`. While previously it returned 0.

Additionally, this is causing the crash that you saw. When `_dynamicSectionIndex` is null, it means that no data has been added to the collection view (in the case `useDynamics==true`). So the collection view datasource is returning x number of rows for section 0 when no one has been inserted.

Next code should be fix the crash:
```swift
    open func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        if hasHeader() && section == 0 { return 0 }
        let rows = _sections[actionSectionIndexFor(section)].actions.count

        guard let dynamicSectionIndex = _dynamicSectionIndex else {
            return settings.behavior.useDynamics ? 0 : rows
        }
        if settings.behavior.useDynamics && section > dynamicSectionIndex {
            return 0
        }
        return rows
    }
```

m-revetria commented on Source/ActionController.swift:
```diff
@@ -265,340 +266,340 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
         let action = self.actionForIndexPath(actionIndexPathFor(indexPath))
 
-        if let action = action where action.executeImmediatelyOnTouch {
+        if let action = action , action.executeImmediatelyOnTouch {
```
please remove the white space before the comma

m-revetria commented on Example/CustomActionControllers/Twitter.swift:
```diff
@@ -98,59 +98,63 @@ public class TwitterActionControllerHeader: UICollectionReusableView {
         }
     }
+  
+  required public init?(coder aDecoder: NSCoder) {
```
please indent using 4 spaces

m-revetria commented on Example/CustomActionControllers/Youtube.swift:
```diff
@@ -91,16 +91,20 @@ public class YoutubeActionController: ActionController<YoutubeCell, ActionData,
             cell.alpha = action.enabled ? 1.0 : 0.5
             
-            UIView.animateWithDuration(0.30) {
+          UIView.animate(withDuration: 0.30) {
```
please fix indentation

mohpor replied:
Sure, But I fail to see the point of an empty animation block! Did you mean to put that alpha value change inside the animation blog?

m-revetria commented on Source/ActionController.swift:
```diff
@@ -179,84 +179,85 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
         // calculate content Inset
         collectionView.layoutSubviews()
-        if let section = _sections.last where !settings.behavior.useDynamics {
+        if let section = _sections.last , !settings.behavior.useDynamics {
```
please delete white space before comma

mohpor replied:
This is caused by the Xcode migrator tool which replaced `where` with `,` (Notice that there is a whitespace before and after `where` but there shouldn't be white space there. I shouldn't have overlooked these. Sorry)

m-revetria commented on Source/ActionController.swift:
```diff
@@ -265,340 +266,340 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
     }
     
-    public override func prefersStatusBarHidden() -> Bool {
+    open override var prefersStatusBarHidden : Bool {
```
please delete white space before colon

m-revetria commented on Source/ActionController.swift:
```diff
@@ -265,340 +266,340 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
     }
     
-    public  override func preferredStatusBarStyle() -> UIStatusBarStyle {
+    open  override var preferredStatusBarStyle : UIStatusBarStyle {
```
here too and place remove extra space between `open` and `override`

m-revetria commented on Source/ActionController.swift:
```diff
@@ -265,340 +266,340 @@ public class ActionController<ActionViewType: UICollectionViewCell, ActionDataTy
 
         self.dismiss() {
-            if let action = action where !action.executeImmediatelyOnTouch {
+            if let action = action , !action.executeImmediatelyOnTouch {
```
please remove space before comma

8masterofpuppets8 commented:
Just noticed this PR and was wondering if it will get merged in at some point? This is the relevant issue for this #25

ryanholden8 commented:
Looking forward to the merge, great work!

thonydam commented:
👍

mohpor commented:
@m-revetria, I have rebased the PR based on your comments.
The error has been fixed for TwitterBot example thanks to the code you've provided.
(Honestly, I have converted so many Libraries to Swift 3 in the past couple of weeks, my mind's playing tricks on me and makes me do stupid things like coercing nulls into 0! sorry about that)

m-revetria commented:
@mohpor great work! thanks for your help and your time updating this PR!
